### ISSUE TITLE
test-execute fails on Fedora 36

### ISSUE THREAD

**Issue by daandemeyer** [systemd/systemd#23332]
Relevant output:

```
exec-umask-namespace.service: Will spawn child (service_enter_start): /bin/ls
exec-umask-namespace.service: Passing 0 fds to service
exec-umask-namespace.service: About to execute /bin/ls -lahd /tmp/subdir
exec-umask-namespace.service: Forked /bin/ls as 297773
exec-umask-namespace.service: Changed dead -> start
exec-umask-namespace.service
        UMask: 0007
        WorkingDirectory: /
        RootDirectory: /
        NonBlocking: no
        PrivateTmp: no
        PrivateDevices: no
        ProtectKernelTunables: no
        ProtectKernelModules: no
        ProtectKernelLogs: no
        ProtectClock: no
        ProtectControlGroups: no
        PrivateNetwork: no
        PrivateUsers: no
        ProtectHome: no
        ProtectSystem: no
        MountAPIVFS: no
        IgnoreSIGPIPE: yes
        MemoryDenyWriteExecute: no
        RestrictRealtime: no
        RestrictSUIDSGID: no
        KeyringMode: private
        ProtectHostname: no
        ProtectProc: default
        ProcSubset: all
        RuntimeDirectoryPreserve: no
        RuntimeDirectoryMode: 0755
        StateDirectoryMode: 0755
        CacheDirectoryMode: 0755
        LogsDirectoryMode: 0755
        ConfigurationDirectoryMode: 0755
        TimeoutCleanSec: infinity
        StandardInput: null
        StandardOutput: null
        StandardError: inherit
        User: 65534
        Group: 65534
        DynamicUser: no
        BindPaths: /etc:/tmp/subdir/subsub:rbind
        TemporaryFileSystem: /tmp:ro
        LockPersonality: no
        SystemCallErrorNumber: kill
exec-umask-namespace.service: User lookup succeeded: uid=65534 gid=65534
Received SIGCHLD from PID 297773 ((ls)).
Child 297773 ((ls)) died (code=exited, status=226/NAMESPACE)
exec-umask-namespace.service: Child 297773 belongs to exec-umask-namespace.service.
exec-umask-namespace.service: Main process exited, code=exited, status=226/NAMESPACE
exec-umask-namespace.service: Failed with result 'exit-code'.
exec-umask-namespace.service: Service will not restart (restart setting)
exec-umask-namespace.service: Changed start -> failed
exec-umask-namespace.service: Failed to delete cgroup entry from LSM BPF map: No such file or directory
exec-umask-namespace.service: Unit entered failed state.
        PID: 297773
        Start Timestamp: Tue 2022-05-10 10:54:57 CEST
        Exit Timestamp: Tue 2022-05-10 10:54:57 CEST
        Exit Code: exited
        Exit Status: 226
```

**Comment by keszybz** [comment]
Hmm, I can't reproduce this here on a standard F36 installation with updates-testing. What is your environment?

**Comment by DaanDeMeyer** [comment]
Good question, It's the work laptop so there's probably a bunch of non-standard stuff going on.

You can try in mkosi with Fedora 36, that'll trigger it as well.

**Comment by anitazha** [comment]
I was able to trigger it a few times on a fresh F36 VM but now it doesn't repro anymore. It does consistently repro on mkosi.

Through mkosi I noticed that if I remove the test that triggers exec-specifier-credentials-dir.service, or reorder exec-umask-namespace.service to come before it in test-execute, it appears to consistently pass. So I wonder if LoadCredentials= is interfering with with the test-execute unit test environment somehow.

**Comment by yuwata** [comment]
@anitazha and @DaanDeMeyer Could you provide the reproducer? I cannot reproduce it.

BTW, @DaanDeMeyer the failure exec-umask-namespace.service is really unexpected? The second trial of the service should be failed with 226/NAMESPACE. Please find the log generated from the lines below:
https://github.com/systemd/systemd/blob/b380b6438361e39ad5076bfa58d2021621a957af/src/test/test-execute.c#L100-L105

**Comment by DaanDeMeyer** [comment]
@yuwata It's definitely the first trial of the service that fails

Like Anita mentioned, it seems you have to run all the tests for the failure to trigger. If I run `sudo build/test-execute`, it doesn't fail. If I run `sudo ninja -C build test`, it does fail.

Running in mkosi also still fails: `sudo mkosi -d fedora -ff` should trigger the failure.

I'll check on my Arch laptop as well to see if the host machine somehow influences the mkosi tests.

**Comment by DaanDeMeyer** [comment]
I can't reproduce on an Arch machine directly.

I can reproduce the failure in a Fedora container running on an Arch host (`sudo mkosi -d fedora -ff`).

**Comment by yuwata** [comment]
Indeed.
```
src/test/test-execute.c:1110:test_exec_umask_namespace: exec-umask-namespace.service: can_unshare=yes: exit status 226, expected 0
```

**Comment by yuwata** [comment]
With some additional logs, I found /run/systemd/inaccessible/dir is not created.
```
Bind-mounting / on /run/systemd/unit-root (MS_BIND|MS_REC "")...
Applying namespace mount on /run/systemd/unit-root/run/credentials
Bind-mounting /run/systemd/inaccessible/dir on /run/systemd/unit-root/run/credentials (MS_BIND|MS_REC "")...
Failed to mount '/run/systemd/inaccessible/dir' on '/proc/self/fd/4': No such file or directory
Failed to mount /run/systemd/inaccessible/dir (type n/a) on /run/systemd/unit-root/run/credentials (MS_BIND|MS_REC ""): No such file or directory
lstat(/run/systemd/unit-root/run/credentials) succeeds. type=directory, mode=0755
stat(/run/systemd/unit-root/run/credentials) succeeds. type=directory, mode=0755
Failed to lstat '/run/systemd/inaccessible/dir': No such file or directory
Failed to mount /run/systemd/inaccessible/dir to /run/systemd/unit-root/run/credentials: No such file or directory
```

**Comment by yuwata** [comment]
@DaanDeMeyer and @anitazha Could you test #23385? I tested it works on mkosi. But I am not sure why such fix is necessary.

**Comment by DaanDeMeyer** [comment]
That does fix the issue, but it feels like a workaround since the test does succeed when it's executed individually. Maybe we should try to find the test that's interfering with test-execute since the failure only happens when executing all tests?

**Comment by anitazha** [comment]
> Maybe we should try to find the test that's interfering with test-execute since the failure only happens when executing all tests?

It's specifically due to `test(m, "exec-specifier-credentials-dir.service", 0, CLD_EXITED);` in `test_exec_specifier` in test-execute: https://github.com/systemd/systemd/blob/72db29beb50c55021d84c0b99b59643d7f43a084/src/test/test-execute.c#L1083

Like I mentioned above if you remove this test or reorder in this array, `test_exec_umask_namespace` will fail: https://github.com/systemd/systemd/blob/72db29beb50c55021d84c0b99b59643d7f43a084/src/test/test-execute.c#L1192-L1195

I think Yu is on the right track as `test_exec_umask_namespace` seems to be failing to mounting `/run/credentials` as inaccessible as part of the namespacing setup. If I change https://github.com/systemd/systemd/blob/72db29beb50c55021d84c0b99b59643d7f43a084/src/core/namespace.c#L2325-L2326 to READONLY it surprisingly passes (not a fix BTW since logically it's wrong to do so). But I don't know why. The service manager for both tests should be the same so I wouldn't think there would be a permissions issue.

**Comment by DaanDeMeyer** [comment]
> It's specifically due to test(m, "exec-specifier-credentials-dir.service", 0, CLD_EXITED); in test_exec_specifier in test-execute:

But if a test in the same binary is the root cause, why does the failure only trigger when running all tests and not when running only test-execute? I figured the root cause would be a test in a different binary or maybe something in the meson environment.

**Comment by anitazha** [comment]
> But if a test in the same binary is the root cause, why does the failure only trigger when running all tests and not when running only test-execute?

I modified mkosi.build to only run test-execute and it still fails with the same error. test-execute is unique in that it creates a new service manager to run the units. But it also has some reliance on the host environment. So to me it does make some sense that another test in test-execute using a test service manager could break another test using the same test service manager.

**Comment by yuwata** [comment]
> > But if a test in the same binary is the root cause, why does the failure only trigger when running all tests and not when running only test-execute?
> 
> I modified mkosi.build to only run test-execute and it still fails with the same error. test-execute is unique in that it creates a new service manager to run the units. But it also has some reliance on the host environment. So to me it does make some sense that another test in test-execute using a test service manager could break another test using the same test service manager.

Yeah, I also confirm that. So, the remaining questions are
1. Why /run/systemd does not exist when running `ninja test`? There are two possibilities:
  - systemd is really not running, though I am not sure as I am not familiar with what mkosi does.
  - systemd is running, but someone removes it.
2. Is it really reproducible without mkosi? Especially on the system whose PID1 is systemd.

The analysis by @anitazha in the above is correct. We have a test for `InaccessiblePath=`, but it is conditionalized with `is_inaccessible_available()`. But the failed test is not conditionalized, and exec-specifier-credentials-dir.service triggers to enable implicit `InaccessiblePath=` for all later services with mount namespace.

**Comment by yuwata** [comment]
I opened another PR #23405. PTAL.

**Comment by DaanDeMeyer** [comment]
When we run the tests in mkosi, systemd is not running, the script is executed in the container environment using nspawn with a minimal pid1 and not full-blown systemd.

> But the failed test is not conditionalized, and exec-specifier-credentials-dir.service triggers to enable implicit InaccessiblePath= for all later services with mount namespace.

Isn't this what we should fix? Having a test that implicitly affects all further tests that are executed doesn't seem ideal. A quick fix could be to put the credentials test last so that it can't affect any further tests.

**Comment by yuwata** [comment]
> When we run the tests in mkosi, systemd is not running, the script is executed in the container environment using nspawn with a minimal pid1 and not full-blown systemd.

Thanks.

> > But the failed test is not conditionalized, and exec-specifier-credentials-dir.service triggers to enable implicit InaccessiblePath= for all later services with mount namespace.
> 
> Isn't this what we should fix? Having a test that implicitly affects all further tests that are executed doesn't seem ideal. A quick fix could be to put the credentials test last so that it can't affect any further tests.

In test-execute, we have only one service that explicitly has `InaccessiblePath=`, but many services implicitly have it. And all of them, except for exec-umask-namespace.service, are conditionalized with `is_inaccessible_available()`. Hence, I think the correct way to fix the issue is one of #23385 or #23405.
